advection_test/functions.py

```diff
@@ -0,0 +1,1302 @@
+
+
+def run_advectionStepWithReset(
+    tracAdv_dataInit, tracAdv_data, tracAdv, dt, mpi_rank=None
+):
+    """
+    Use: tracAdv_data = run_advectionStepWithReset()
+
+    Inputs:
+    - tracAdv_dataInit: Dict{'tracers', 'delp', 'mfxd', 'mfyd', 'crx', 'cry'}
+                        of values to be reset to after each advection step
+    - tracAdv_data: Dict{'tracers', 'delp', 'mfxd', 'mfyd', 'crx', 'cry'}
+    - tracAdv: class instance of TracerAdvection
+    - dt: time step in seconds
+    - mpi_rank: if 0, prints differences from timestep, initial condition
+
+    Outputs:
+    - tracAdv_data: updated fields (of only tracer)
+    """
+
+    tmp = cp.deepcopy(tracAdv_data["tracers"])  # pre-advection tracer state
+
+    tracAdv(
+        tracAdv_data["tracers"],
+        tracAdv_data["delp"],
+        tracAdv_data["mfxd"],
+        tracAdv_data["mfyd"],
+        tracAdv_data["crx"],
+        tracAdv_data["cry"],
+        dt,
+    )
+
+    tracAdv_data["delp"] = cp.deepcopy(tracAdv_dataInit["delp"])
+    tracAdv_data["mfxd"] = cp.deepcopy(tracAdv_dataInit["mfxd"])
+    tracAdv_data["mfyd"] = cp.deepcopy(tracAdv_dataInit["mfyd"])
+    tracAdv_data["crx"] = cp.deepcopy(tracAdv_dataInit["crx"])
+    tracAdv_data["cry"] = cp.deepcopy(tracAdv_dataInit["cry"])
+
+    if mpi_rank == 0:
+        diff_timestep = tracAdv_data["tracers"]["smoke"].data - tmp["smoke"].data
+        diff_fromInit = (
+            tracAdv_data["tracers"]["smoke"].data
+            - tracAdv_dataInit["tracers"]["smoke"].data
+        )
+        print(
+            "timestep diff min=%.2e; max=%.2e; from init diff min=%.2e; max=%.2e"
+            % (
+                np.nanmin(diff_timestep),
+                np.nanmax(diff_timestep),
+                np.nanmin(diff_fromInit),
+                np.nanmax(diff_fromInit),
+            )
+        )
+
+    return tracAdv_data
+
+
+def plot_projection_field(
+    lon, lat, field, cmap="viridis", vmin=-1, vmax=1, units="", title="", fSave=None
+):
+    """
+    Use: plot_projection_field(
+        lon, lat, field, cmap='viridis', vmin=-1, vmax=1, units='', title='')
+
+    Creates a Robinson projection and plots the (6) tiles on a map.
+
+    Inputs:
+    - lon, lat: lon and lat of coordinate edges (tile, x, y)
+    - field: unstaggered field at a given vertical level (tile, x, y)
+    - cmap: colormap
+    - vmin, vmax: limits of the color map
+    - units: label the units on the color bar
+    - title: set title of the plot
+
+    Outputs: none
+    """
+
+    fig = plt.figure(figsize=(8, 4))
+    fig.patch.set_facecolor("white")
+    ax = fig.add_subplot(111, projection=ccrs.Robinson())
+    ax.set_facecolor(".4")
+
+    f1 = pcolormesh_cube(lat, lon, field, cmap=cmap, vmin=vmin, vmax=vmax)
+    plt.colorbar(f1, label=units)
+
+    ax.set_title(title)
+
+    if fSave is not None:
+        plt.savefig(fSave, dpi=200, bbox_inches="tight")
+        plt.close("all")
+    else:
+        plt.show()
+
+    return
```

Comment:
what is the point of empty return statements at the end of functions? This happens implicitly